codegen/src/test/kotlin/software/amazon/smithy/rust/codegen/smithy/EndpointTraitBindingsTest.kt

```diff
@@ -0,0 +1,142 @@
+        namespace test
+
+        @readonly
+        @endpoint(hostPrefix: "{foo}.data")
+        operation GetStatus {
+            input: GetStatusInput,
+        }
+
+        structure GetStatusInput {
+            @required
+            @hostLabel
+            foo: String
+        }
+        """.asSmithyModel()
+        val operationShape: OperationShape = model.lookup("test#GetStatus")
+        val sym = testSymbolProvider(model)
+        val endpointBindingGenerator = EndpointTraitBindings(
+            model,
+            sym,
+            operationShape,
+            operationShape.expectTrait(EndpointTrait::class.java)
+        )
+        val project = TestWorkspace.testProject()
+        project.withModule(RustModule.default("test", false)) {
+            it.rust(
+                """
+                struct GetStatusInput {
+                    foo: String
+                }
+            """
+            )
+            it.implBlock(model.lookup("test#GetStatusInput"), sym) {
+                it.rustBlock(
+                    "fn endpoint_prefix(&self) -> #T::endpoint::EndpointPrefix",
+                    TestRuntimeConfig.smithyHttp()
+                ) {
+                    endpointBindingGenerator.render(this, "self")
+                    rust(".unwrap()")
+                }
+            }
+            it.unitTest(
+                """
+                let inp = GetStatusInput { foo: "test_value".to_string() };
+                let prefix = inp.endpoint_prefix();
+                assert_eq!(prefix.as_str(), "test_value.data");
+            """
+            )
+        }
+
+        project.compileAndTest()
+    }
+
+    @Test
+    fun `endpoint integration test`() {
+        val model = """
+        namespace com.example
+
+        use aws.protocols#awsJson1_0
+
+        @awsJson1_0
+        @aws.api#service(sdkId: "Test", endpointPrefix: "differentprefix")
+        service TestService {
+            operations: [SayHello],
+            version: "1"
+        }
+
+        @endpoint(hostPrefix: "test123.{greeting}.")
+        operation SayHello {
+            input: SayHelloInput
+        }
+
+        structure SayHelloInput {
+            @required
+            @hostLabel
+            greeting: String
+        }
+        """.asSmithyModel()
+        val (ctx, testDir) = generatePluginContext(model, "com.example#TestService")
+        val testDecorator = object : RustCodegenDecorator {
+            override val name: String = "Test"
+            override val order: Byte = 0
+            override fun extras(protocolConfig: ProtocolConfig, rustCrate: RustCrate) {
+                rustCrate.withModule(RustModule.default("test", public = false)) {
+                    it.unitTest(
+                        """
+                let conf = crate::Config::builder().build();
+                crate::operation::SayHello::builder().greeting("hey there!").build(&conf).expect_err("no spaces or exclamation points in ep prefixes");
```

Comment:
Should there also be a test for empty host labels (from the Client Behavior section of the spec)?

> Clients MUST fail when expanding a hostPrefix template if the value of any labeled member is empty or null.

Reference: https://awslabs.github.io/smithy/1.0/spec/core/endpoint-traits.html#client-behavior